gherkin/php/src/GherkinDocumentBuilder.php

```diff
@@ -0,0 +1,371 @@
+                    tags: $tags,
+                    language: $language,
+                    keyword: $featureLine->matchedKeyword ?? '',
+                    name: $featureLine->matchedText ?? '',
+                    description: $this->getDescription($header),
+                    children: $children,
+                );
+
+            case RuleType::Rule:
+
+                $header = $node->getSingle(AstNode::class, RuleType::RuleHeader, new AstNode(RuleType::RuleHeader));
+
+                $ruleLine = $header->getToken(TokenType::RuleLine);
+
+                $children = [];
+                $tags = $this->getTags($header);
+
+                $background = $node->getSingle(Background::class, RuleType::Background);
+                if ($background) {
+                    $children[] = new RuleChild(background: $background);
+                }
+                $scenarios = $node->getItems(Scenario::class, RuleType::ScenarioDefinition);
+                foreach ($scenarios as $scenario) {
+                    $children[] = new RuleChild(scenario: $scenario);
+                }
+
+                return new Rule(
+                    location: $this->getLocation($ruleLine, 0),
+                    tags: $tags,
+                    keyword: $ruleLine->matchedKeyword ?? '',
+                    name: $ruleLine->matchedText ?? '',
+                    description: $this->getDescription($header),
+                    children: $children,
+                    id: $this->idGenerator->newId(),
+                );
+
+            case RuleType::GherkinDocument:
+                $feature = $node->getSingle(Feature::class, RuleType::Feature);
+
+                return new GherkinDocument(
+                    uri: $this->uri,
+                    feature: $feature,
+                    comments: $this->comments,
+                );
+
+            default:
+                return $node;
+        }
+    }
+
+    private function getLocation(Token $token, int $column): MessageLocation
+    {
+        $column = ($column === 0) ? $token->location->column : $column;
+
+        return new MessageLocation($token->location->line, $column);
+    }
+
+    private function getDescription(AstNode $node): string
+    {
+        return (string) $node->getSingleUntyped(RuleType::Description, "");
+    }
+
+    /** @return list<Step> */
+    private function getSteps(AstNode $node): array
+    {
+        return $node->getitems(Step::class, RuleType::Step);
+    }
+
+    /** @return list<TableRow> */
+    private function getTableRows(AstNode $node): array
+    {
+        $rows = [];
+
+        foreach ($node->getTokens(TokenType::TableRow) as $token) {
+            $rows[] = new TableRow($this->getLocation($token, 0), $this->getCells($token), $this->idGenerator->newId());
+        }
+
+        $this->ensureCellCount($rows);
+
+        return $rows;
+    }
+
+    /** @param list<TableRow> $rows */
+    private function ensureCellCount(array $rows): void
```

Comment:
let's make those methods static?

Reply:
I don't think I will here, I'm not 100% on the pattern (can change later)